### ISSUE TITLE
single_element_loop: syntax error with `never patterns` ( missing parens)

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#12782]
### Summary

.

### Lint Name

single_element_loop

### Reproducer

I tried this code:

```rust
#![feature(never_patterns)]
#![allow(incomplete_features)]
#![warn(clippy::single_element_loop)]
// edition: 2021

#[derive(Copy, Clone)]
enum Void {}

fn main() {
    let res_void: Result<bool, Void> = Ok(true);

    for (Ok(mut _x) | Err(!)) in [res_void] {
    // sugg:
    //{ let Ok(mut _x) | Err(!) = res_void;
       let ptr: *const Void = std::ptr::null();
    }
}

```

I saw this happen:
suggestion does not build:
```
error: top-level or-patterns are not allowed in `let` bindings
  --> seloop.rs:11:11
   |
11 |     { let Ok(mut _x) | Err(!) = res_void;
   |           ^^^^^^^^^^^^^^^^^^^ help: wrap the pattern in parentheses: `(Ok(mut _x) | Err(!))`
```

### Version

```text
rustc 1.80.0-nightly (faefc618c 2024-05-07)
binary: rustc
commit-hash: faefc618cf48bd794cbc808448df1bf3f59f36af
commit-date: 2024-05-07
host: x86_64-unknown-linux-gnu
release: 1.80.0-nightly
LLVM version: 18.1.4
```

**Comment by GuillaumeGomez** [comment]
There are actually issues, the first one is that the parens are not kept, so this code:

```rust
let res_void: Result<bool, bool> = Ok(true);

for (Ok(mut _x) | Err(mut _x)) in [res_void] {
   let ptr: *const bool = std::ptr::null();
}
```

suggests:

```rust
{
    let Ok(mut _x) | Err(mut _x) = res_void;
    let ptr: *const bool = std::ptr::null();
}
```

Which is wrong:

```
error: top-level or-patterns are not allowed in `let` bindings
 --> src/main.rs:5:9
  |
5 |     let Ok(mut _x) | Err(mut _x) = res_void;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^ help: wrap the pattern in parentheses: `(Ok(mut _x) | Err(mut _x))`
```
